Apache Arrow: Submission

essandess opened the pull request:
Apache Arrow

* Port apache-arrow: C++ libraries and headers
* Ports py*-pyarrow Python bindings
* Update and Bugfixes of py-apache-beam
* Submission of dependency py-pytest-lazy-fixture

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk -v ORS=' ' '{print $NF}')"
-->
macOS 10.x
Xcode 8.x

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+                    size    5835020
+
+    # Apache Arrow requires a python3 version to build; use this
+    set python_branch   3.7
```
Please use 3.8; it's the latest stable version.

essandess replied:
Thanks done.

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+    build.env-append \
+                    "PATH=${python.prefix}/bin:$env(PATH)" \
+                    PYARROW_CMAKE_OPTIONS=${configure.pre_args} \
+                    PYARROW_WITH_PARQUET=1    
```
There is trailing whitespace on this line.

essandess replied:
Thanks fixed.

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+                    "PATH=${python_prefix}/bin:$env(PATH)"
+    destroot.dir    ${build_dir}
+    destroot.cmd    ${build.cmd}
+    destroot.destdir    DESTDIR=${destroot}
```
Oh, maybe not, because you're using the python portgroup here.

essandess replied:
Yep.

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+    destroot.env-append \
+                    "PATH=${python.prefix}/bin:$env(PATH)"
+    destroot.dir    ${build.dir}
+    destroot.target install
```
Oh, maybe not, because you're using the python portgroup here.

essandess replied:
Yep.

ryandesign commented:
> Port py37-pyarrow **fail** on xcode9.4. [Log](https://paste.macports.org/b0bd59ecc8f0)

Here, the error is:

```
/opt/local/var/macports/build/_Users_travis_build_macports_macports-ports_devel_apache-arrow/apache-arrow/work/apache-arrow-0.17.1/cpp/src/arrow/scalar.cc:424:12: error: call to 'CastImpl' is ambiguous
    return CastImpl(checked_cast<const typename TypeTraits<FromType>::ScalarType&>(from_),
           ^~~~~~~~
```

essandess commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+                    fast.
+
+homepage            https://github.com/apache/arrow
+distname            ${github.author}-${github.project}-${version}
```
Thanks fixed.

essandess commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+distname            ${github.author}-${github.project}-${version}
+
+# Build instructions: https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#build-and-test
+set build_dir       ${worksrcpath}/cpp/build
```
Thanks fixed and improved.

essandess commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+cmake.build_type    Release
+
+configure.pre_args-append \
+                    -DARROW_FLIGHT=ON \
```
Thanks fixed.

essandess commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,164 @@
+
+    build.env-append \
+                    "PATH=${python.prefix}/bin:$env(PATH)" \
+                    PYARROW_CMAKE_OPTIONS=${configure.pre_args} \
```
Awesome thanks for the fix.

This is need as well as fixing the parts where the port groups cmake and python conflict. A few `-replace` commands are all that's needed to set things right.

essandess commented:
> > Port py37-pyarrow **fail** on xcode9.4. [Log](https://paste.macports.org/b0bd59ecc8f0)
> 
> Here, the error is:
> 
> ```
> /opt/local/var/macports/build/_Users_travis_build_macports_macports-ports_devel_apache-arrow/apache-arrow/work/apache-arrow-0.17.1/cpp/src/arrow/scalar.cc:424:12: error: call to 'CastImpl' is ambiguous
>     return CastImpl(checked_cast<const typename TypeTraits<FromType>::ScalarType&>(from_),
>            ^~~~~~~~
> ```

Thanks. That's a consequence of port group python options beings used as cmake flags, which should be fixed  now.

Also, for posterity, errors like this:
```
CMake Error at /opt/local/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find Python3 (missing: Python3_EXECUTABLE
  Python3_NumPy_INCLUDE_DIRS Interpreter NumPy) (found version "3.7.8")
```
seemed to mean that the cmake system found the wrong python executable, e.g.. `python3.8` instead of `python3.7`. I tried several obvious thongs to fix like setting `-DPYTHON3_EXECUTABLE` etc. but nothing worked. I just altered `PATH` to make sure it found the correct version. I suspect that this is a bug in upstream's cmake configuration.

essandess commented:
Re libraries, it would be nice to have a command like this saved somewhere in macports. For posterity:
```
port contents apache-arrow | egrep -e '\.dylib$' | xargs otool -L | egrep -o -e '/opt/local/.*\.dylib' | uniq | xargs port provides | sed -E -e 's/^.+by: //g' | sort | uniq
```

> apache-arrow links with all sorts of libraries that it doesn't declare dependencies on:
> 
> ```
> 	/opt/local/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.8)
> 	/opt/local/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
> 	/opt/local/lib/liblz4.1.dylib (compatibility version 1.0.0, current version 1.9.2)
> 	/opt/local/lib/libprotobuf.23.dylib (compatibility version 24.0.0, current version 24.3.0)
> 	/opt/local/lib/libsnappy.1.dylib (compatibility version 1.0.0, current version 1.1.8)
> 	/opt/local/lib/libssl.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
> 	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
> 	/opt/local/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.4.5)
> ```

essandess commented:
I get a successful build and test on my system. The error in the builedbot suggests that the cmake system can't find the `apache-arrow` install. Not sure why:

https://paste.z0k.xyz/a953d943d11c.txt
```
-- Running cmake for pyarrow
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/local -DCMAKE_INSTALL_NAME_DIR=/opt/local/lib -DCMAKE_SYSTEM_PREFIX_PATH=/opt/local;/usr -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_POLICY_DEFAULT_CMP0060=NEW -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_COLOR_MAKEFILE=ON -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_MAKE_PROGRAM=/usr/bin/make -DCMAKE_MODULE_PATH=/opt/local/share/cmake/Modules -DCMAKE_PREFIX_PATH=/opt/local/share/cmake/Modules -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_RPATH=/opt/local/lib -Wno-dev -DARROW_FLIGHT=ON -DARROW_JEMALLOC=OFF -DARROW_ORC=ON -DARROW_PARQUET=ON -DARROW_PLASMA=ON -DARROW_PROTOBUF_USE_SHARED=ON -DARROW_PYTHON=ON -DARROW_WITH_BZ2=ON -DARROW_WITH_ZLIB=ON -DARROW_WITH_ZSTD=ON -DARROW_WITH_LZ4=ON -DARROW_WITH_SNAPPY=ON -DARROW_WITH_BROTLI=ON -DARROW_INSTALL_NAME_RPATH=OFF -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8  -DPYARROW_BUILD_CUDA=off -DPYARROW_BUILD_FLIGHT=off -DPYARROW_BUILD_GANDIVA=off -DPYARROW_BUILD_DATASET=off -DPYARROW_BUILD_ORC=off -DPYARROW_BUILD_PARQUET=on -DPYARROW_BUILD_PLASMA=off -DPYARROW_BUILD_S3=off -DPYARROW_BUILD_HDFS=off -DPYARROW_USE_TENSORFLOW=off -DPYARROW_BUNDLE_ARROW_CPP=off -DPYARROW_BUNDLE_BOOST=off -DPYARROW_GENERATE_COVERAGE=off -DPYARROW_BOOST_USE_SHARED=on -DPYARROW_PARQUET_USE_SHARED=on -DCMAKE_BUILD_TYPE=release /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py38-pyarrow/work/arrow-0.17.1/python
…
-- Found Arrow: /opt/local/include (found version "0.17.1")
-- Arrow version: 0.17.1 (CMake package configuration: Arrow)
-- Arrow SO and ABI version: 17
-- Arrow full SO version: 17.1.0
-- Found the Arrow core shared library: /opt/local/lib/libarrow.17.1.0.dylib
-- Found the Arrow core import library:
-- Found the Arrow core static library: /opt/local/lib/libarrow.a
-- Could NOT find ArrowPython (missing: ArrowPython_DIR)
CMake Error at /opt/local/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find ArrowPython (missing: ARROW_PYTHON_INCLUDE_DIR
  ARROW_PYTHON_LIB_DIR) (found version "0.17.1")
Call Stack (most recent call first):
  /opt/local/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
  cmake_modules/FindArrowPython.cmake:76 (find_package_handle_standard_args)
  CMakeLists.txt:210 (find_package)
```

essandess commented:
This trace gives me the same error on my system:
```
sudo bash -c 'port uninstall py38-pyarrow ; port clean --all apache-arrow py38-pyarrow ; port -pNvst install py38-pyarrow'
```

I don't see the cause.

Why would `port install` work, but not `port -vst install` in this case?

essandess commented:
One of the azure build logs showed that `py27-pyarrow` built successfully, so I tried restricting everything to python 37, but that doesn't work either. Per discussion over dev@arrow.apache.org, I'm opening a JIRA ticket at https://issues.apache.org/jira/projects/ARROW/issues.

I'm still scratching my head why this works on a local install in non-trace mode, but fails in trace or on azure.

essandess commented:
The build results are very inconsistent.
* Successful build of `py37-pyarrow`: https://paste.z0k.xyz/d01a4e4e9854.txt
* Unsuccessful build of `py38-pyarrow`: https://paste.z0k.xyz/0d4396087cd0.txt

kencu commented:
> The build results are very inconsistent.
> 
> * Successful build of `py37-pyarrow`: https://paste.z0k.xyz/d01a4e4e9854.txt
> * Unsuccessful build of `py38-pyarrow`: https://paste.z0k.xyz/0d4396087cd0.txt

py37* finds pkg-config, py38* does not. Why?

essandess commented:
@kencu @ryandesign 

Thanks, I think I've addressed the issues above by adding dependencies and can now do a successful local `port -pNvst install py38pyarrow`.

However, now I'm hitting an azure error with the wrong version `/opt/local/lib/libprotoc.22.dylib` being used in `/opt/local/bin/grpc_cpp_plugin`.

My local version has `/opt/local/lib/libprotoc.23.dylib` .

Please see: https://paste.z0k.xyz/565bee7c028f.txt

This looks like a build bug. Do you agree?

In the meantime, I can just try the build without the `grpc` dependency.

essandess commented:
It is a bug. I'll just update to latest and hope it goes away…
```
$ sudo port install grpc
--->  Computing dependencies for grpc
--->  Fetching archive for grpc
--->  Attempting to fetch grpc-1.22.0_2.darwin_19.x86_64.tbz2 from https://packages.macports.org/grpc
--->  Attempting to fetch grpc-1.22.0_2.darwin_19.x86_64.tbz2.rmd160 from https://packages.macports.org/grpc
--->  Installing grpc @1.22.0_2
--->  Activating grpc @1.22.0_2
--->  Cleaning grpc
--->  Scanning binaries for linking errors
--->  Found 20 broken files, matching files to ports     
--->  Found 1 broken port, determining rebuild order
You can always run 'port rev-upgrade' again to fix errors.
The following ports will be rebuilt: grpc @1.22.0
Continue? [Y/n]: n
$ otool -L /opt/local/bin/grpc_cpp_plugin 
/opt/local/bin/grpc_cpp_plugin:
	/opt/local/lib/libprotoc.22.dylib (compatibility version 23.0.0, current version 23.4.0)
…
```

kencu commented on devel/grpc/Portfile:
```diff
@@ -1,53 +1,76 @@
+maintainers         nomaintainer
+license             Apache-2
+fetch.type          git
 
```
You don't usually need this if you use the github PG -- is there something weird about this port to need it?

essandess commented:
@kencu @ryandesign This Portfile for `py-apache-beam` still has a subtle bug:

For some reason, the installed `PKG-INFO` and other stuff in the egg directory is not readable by others, and this prevents imports:

```bash
$ ls -l /opt/local/ports/python/py-apache-beam/work/destroot/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/apache_beam-2.22.0-py3.7.egg-info/PKG-INFO
-rw-r-----  1 root  wheel  1339 Jul 23 20:55 /opt/local/ports/python/py-apache-beam/work/destroot/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/apache_beam-2.22.0-py3.7.egg-info/PKG-INFO
```

Perhaps this is inherited from the same permissions issues on the extracted directory.

Any suggestions on what could cause this issue? I can go change the permissions by hand, but perhaps there's a more general solution.

```bash
$ ls -ld /opt/local/ports/python/py-apache-beam/work/apache-beam-2.22.0/
drwxr-x---  11 macports  wheel  352 Jul 23 20:55 /opt/local/ports/python/py-apache-beam/work/apache-beam-2.22.0//
```

essandess commented:
@kencu @ryandesign To the best of my knowledge this is ready.

wesm commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,209 @@
+
+    # Note: Modifying PATH is required to ensure that cmake finds the correct python version
+    configure.env-append \
+                    "PATH=${python_prefix}/bin:${workpath}/bin:$env(PATH)"
```
Would it be better to use `-DPYTHON_EXECUTABLE=...` when invoking CMake?

essandess replied:
I tried that but it doesn't work. See this issue: https://issues.apache.org/jira/browse/ARROW-9532?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel

essandess replied:
Homebrew has it working, so I'll try it again.

essandess replied:
Thanks, got that working now. I must've fumbled the earlier attempt.

essandess replied:
Nope, `py37-pyarrow` fails to build without modifying path, in spite of specifying `-DPYTHON_EXECUTABLE=` in both the `cmake` arguments and `PYARROW_CMAKE_OPTIONS`.

I believe that this is an issue with the upstream `cmake` configuration.

Here are the error logs for the failed py37-pyarrow build: https://paste.z0k.xyz/b822c7df5064
```
--->  Building py37-pyarrow
DEBUG: Executing proc-pre-org.macports.build-build-0
DEBUG: Executing org.macports.build (py37-pyarrow)
DEBUG: Environment: 
CC='/usr/bin/clang'
CC_PRINT_OPTIONS='YES'
CC_PRINT_OPTIONS_FILE='/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/.CC_PRINT_OPTIONS'
CFLAGS='-arch x86_64 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
CPATH='/opt/local/include'
CXX='/usr/bin/clang++'
CXXFLAGS='-arch x86_64 -stdlib=libc++ -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
DEVELOPER_DIR='/Library/Developer/CommandLineTools'
F90FLAGS='-m64'
FCFLAGS='-m64'
FFLAGS='-m64'
LDFLAGS='-arch x86_64'
LIBRARY_PATH='/opt/local/lib'
MACOSX_DEPLOYMENT_TARGET='10.14'
OBJC='/usr/bin/clang'
OBJCFLAGS='-arch x86_64 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
PATH='/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/bin:/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin'
PYARROW_CMAKE_OPTIONS='-DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX="/opt/local" -DCMAKE_INSTALL_NAME_DIR="/opt/local/lib" -DCMAKE_SYSTEM_PREFIX_PATH="/opt/local;/usr" -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_POLICY_DEFAULT_CMP0060=NEW -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_COLOR_MAKEFILE=ON -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_MAKE_PROGRAM=/usr/bin/make -DCMAKE_MODULE_PATH="/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm" -DCMAKE_PREFIX_PATH="/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm" -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_RPATH="/opt/local/lib" -Wno-dev -DARROW_FLIGHT=ON -DARROW_JEMALLOC=OFF -DARROW_ORC=ON -DARROW_PARQUET=ON -DARROW_PLASMA=ON -DARROW_PROTOBUF_USE_SHARED=ON -DARROW_PYTHON=ON -DARROW_USE_CCACHE=OFF -DARROW_WITH_BZ2=ON -DARROW_WITH_ZLIB=ON -DARROW_WITH_ZSTD=ON -DARROW_WITH_LZ4=ON -DARROW_WITH_SNAPPY=ON -DARROW_WITH_BROTLI=ON -DARROW_INSTALL_NAME_RPATH=OFF -DCARES_PREFIX=/opt/local -DgRPC_INSTALL=OFF -DgRPC_BUILD_TESTS=OFF -DgRPC_CARES_PROVIDER=package -DgRPC_ABSL_PROVIDER=package -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_RE2_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DLLVM_ROOT=/opt/local/libexec/llvm-10 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7'
PYARROW_WITH_PARQUET='1'
SDKROOT='/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
Executing:  cd "/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build_ext --inplace 
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build_ext --inplace 
running build_ext
creating /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build
creating /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build/temp.macosx-10.14-x86_64-3.7
-- Running cmake for pyarrow
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/local -DCMAKE_INSTALL_NAME_DIR=/opt/local/lib -DCMAKE_SYSTEM_PREFIX_PATH=/opt/local;/usr -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_POLICY_DEFAULT_CMP0060=NEW -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_COLOR_MAKEFILE=ON -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_MAKE_PROGRAM=/usr/bin/make -DCMAKE_MODULE_PATH=/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm -DCMAKE_PREFIX_PATH=/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_RPATH=/opt/local/lib -Wno-dev -DARROW_FLIGHT=ON -DARROW_JEMALLOC=OFF -DARROW_ORC=ON -DARROW_PARQUET=ON -DARROW_PLASMA=ON -DARROW_PROTOBUF_USE_SHARED=ON -DARROW_PYTHON=ON -DARROW_USE_CCACHE=OFF -DARROW_WITH_BZ2=ON -DARROW_WITH_ZLIB=ON -DARROW_WITH_ZSTD=ON -DARROW_WITH_LZ4=ON -DARROW_WITH_SNAPPY=ON -DARROW_WITH_BROTLI=ON -DARROW_INSTALL_NAME_RPATH=OFF -DCARES_PREFIX=/opt/local -DgRPC_INSTALL=OFF -DgRPC_BUILD_TESTS=OFF -DgRPC_CARES_PROVIDER=package -DgRPC_ABSL_PROVIDER=package -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_RE2_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DLLVM_ROOT=/opt/local/libexec/llvm-10 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7  -DPYARROW_BUILD_CUDA=off -DPYARROW_BUILD_FLIGHT=off -DPYARROW_BUILD_GANDIVA=off -DPYARROW_BUILD_DATASET=off -DPYARROW_BUILD_ORC=off -DPYARROW_BUILD_PARQUET=on -DPYARROW_BUILD_PLASMA=off -DPYARROW_BUILD_S3=off -DPYARROW_BUILD_HDFS=off -DPYARROW_USE_TENSORFLOW=off -DPYARROW_BUNDLE_ARROW_CPP=off -DPYARROW_BUNDLE_BOOST=off -DPYARROW_GENERATE_COVERAGE=off -DPYARROW_BOOST_USE_SHARED=on -DPYARROW_PARQUET_USE_SHARED=on -DCMAKE_BUILD_TYPE=release /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- System processor: x86_64
-- Performing Test CXX_SUPPORTS_SSE4_2
-- Performing Test CXX_SUPPORTS_SSE4_2 - Success
-- Performing Test CXX_SUPPORTS_AVX2
-- Performing Test CXX_SUPPORTS_AVX2 - Success
-- Performing Test CXX_SUPPORTS_AVX512
-- Performing Test CXX_SUPPORTS_AVX512 - Success
-- Arrow build warning level: PRODUCTION
Configured for RELEASE build (set with cmake -DCMAKE_BUILD_TYPE={release,debug,...})
-- Build Type: RELEASE
-- Build output directory: /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build/temp.macosx-10.14-x86_64-3.7/release
CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  version "3.8.5")
Call Stack (most recent call first):
  /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  /opt/local/share/cmake-3.18/Modules/FindPython/Support.cmake:2971 (find_package_handle_standard_args)
  /opt/local/share/cmake-3.18/Modules/FindPython3.cmake:389 (include)
  cmake_modules/FindPython3Alt.cmake:46 (find_package)
  CMakeLists.txt:196 (find_package)


-- Configuring incomplete, errors occurred!
```

Here's are the logs for a successful build of `py38-pyarrow`: https://paste.z0k.xyz/036eb427e525

essandess replied:
Finally, now it works with `-DPython3_EXECUTABLE=…` thanks to the folks at the arrow jira.

essandess commented:
@kencu @wesm @ryandesign This is ready to merge, and is now a dependency of https://github.com/macports/macports-ports/pull/8783.

Would you please take a look and address this PR?

essandess commented:
There's apparently a cmake issue with Arrow 2.0.0 and the latest grpc version 1.33.1. See https://issues.apache.org/jira/browse/ARROW-10412?jql=project%20%3D%20ARROW%20AND%20resolution%20%3D%20Unresolved%20AND%20text%20~%20%22grpc%22%20ORDER%20BY%20cf%5B12313920%5D%20ASC%2C%20priority%20DESC%2C%20updated%20DESC

ryandesign commented on devel/grpc/Portfile:
```diff
@@ -1,53 +1,114 @@
+build.target
 
-depends_build-append port:pkgconfig
+# fix shared library paths
```
What's wrong with the shared library paths, and could the build system be fixed to generate them correctly in the first place, rather than adding all this difficult-to-understand code to the portfile?

essandess replied:
Here's the issue:

The build links against its own shared libraries in `${cmake.build_dir}`. But these these libraries are built with id `${prefix}/lib`. Because they haven't been installed yet, this causes an "image not found" linker error during the build.

If there's a smarter way to tackle this, I'm not aware of it.

kencu replied:
check this out <https://github.com/macports/macports-ports/blob/d7d881884bbb082a572bb24ddaebcd41fe553e24/_resources/port1.0/group/cmake-1.1.tcl#L105>

essandess replied:
That's the smarter way—thanks!

essandess commented:
@kencu @ryandesign To the best of my knowledge this is robust and ready to merge. Would you please take a look? There are other PRs that have this as a dependency.

ryandesign commented on net/c-ares/Portfile:
```diff
@@ -1,26 +1,32 @@
 
-configure.args   --mandir=${prefix}/share/man \
-                 --enable-shared
+#configure.args   --mandir=${prefix}/share/man \
+#                 --enable-shared
```
Why leave these commented-out lines in the port?

essandess replied:
Oops thanks.

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,223 @@
+                    port:libunwind \
+                    port:lz4 \
+                    port:ninja \
+                    port:openssl \
```
If you know for certain that this port is incompatible with libressl, please add a comment with a URL to the upstream bug report about that.

If you don't know for certain that it's incompatible with libressl, allow libressl to be used by declaring this as a path dependency.

```suggestion
                    path:lib/libssl.dylib:openssl \
```

essandess replied:
Thanks done.

ryandesign commented on devel/apache-arrow/Portfile:
```diff
@@ -0,0 +1,223 @@
+                    port:brotli \
+                    port:bzip2 \
+                    port:c-ares \
+                    port:flatbuffers \
```
The developers of flatbuffers do not believe that it should be part of a package manager like MacPorts; rather, it should be bundled with whatever software wants to use it. See https://github.com/google/flatbuffers/issues/6076#issuecomment-694954041. If and when we remove flatbuffers from MacPorts, it would break this port. Does apache-arrow have a bundled copy of flatbuffers that it could use instead? If not, could you suggest it to them? Also FYI @herbygillot

essandess replied:
Upstream has it: https://github.com/apache/arrow/tree/master/cpp/thirdparty

I verify that `port test` works without MacPorts `flatbuffers`.

essandess commented:
This PR now depends upon the cmake build of `grpc` in https://github.com/macports/macports-ports/pull/9729#event-4211195020

essandess commented:
This PR now depends upon the cmake build of:
* `grpc` #9729
* `c-ares` #9835
* `re2` #9836

kencu commented:
So the gist of this is that Apache Arrow has been written such that it is only capable of using dependencies that have been built with cmake, and have cmake files? 

All other types of packaging info, like pkg-config, are broken and don't work, and Apache Arrow has decided it is cmake or nothing?

That h&^*&^osed.

essandess commented:
Apache Arrow’s build instructions are linked in the Portfile: https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#build-and-test

And this isn’t the only reason for this PR. The port’s old Makefile doesn’t build the latest versions for a while now.

Adding cmake builds for downstream dependencies isn’t a big deal, and provides consistency for other tools that use these too.

> So the gist of this is that Apache Arrow has been written such that it is only capable of using dependencies that have been built with cmake, and have cmake files?
> 
> All other types of packaging info, like pkg-config, are broken and don't work, and Apache Arrow has decided it is cmake or nothing?
> 
> That is h&^*&^osed. Are you certain that is the case? This would rather unique, if so.

essandess commented:
Note that this [force-push](https://github.com/macports/macports-ports/compare/08206525abd6d5644ef62976a288b4a6d859f08c..76927172cfe36acb7a6ea543223dd8849dbb0cb9) builds `apache-arrow`, but exhibits these known issues for `py*-pyarrow`:
* http://mail-archives.apache.org/mod_mbox/arrow-dev/202009.mbox/%3C6ECB7D60-844F-410D-8840-24F2CE261752@yahoo.com%3E
* http://mail-archives.apache.org/mod_mbox/arrow-jira/202009.mbox/%3CJIRA.13327607.1600156125000.372264.1600157460269@Atlassian.JIRA%3E
* https://www.mail-archive.com/dev@arrow.apache.org/msg20631.html

cjones051073 commented:
@essandess Is this port ready to go ? It still seems to be failing to build. Its now a dep. of a number of other submissions so needs to be merged before they can be.

essandess commented:
> @essandess Is this port ready to go ? It still seems to be failing to build. Its now a dep. of a number of other submissions so needs to be merged before they can be.

Yes, and I confirm it’s working locally.

This PR needs these PRs for `re2` with cmake, and an updated `grpc` with cmake along with `py-grpcio` and `py-grpcio-tools`. All are ready and working to the best of my knowledge. The dependency DAG is:

#9836 ← #9729  ← #7791 ← #9840 ← tensorflow stuff ← …

cc: @kencu @reneeotten
